### PR TITLE
Add `groupby.fillna`

### LINKED ISSUE #4896: [FEA] Add forward fill and backward fill options for fillna in groupby objects.
#1361 is an FEA for `df.fillna()`. 

I'm creating this specifically for groupby objects.

### PR DESCRIPTION
**Pull request by isVoid**
Closes #4896

This PR adds support for `groupby.fillna`. Fixes a bug in libcudf `groupby.replace_nulls` when the input is not-nullable.

Example:
```python
>>> df = cudf.DataFrame({'a': [3, 2, 3, 1], 'b': [None, 8, 8, None]}, index=['w', 'x', 'y', 'z'])
>>> df.groupby('a').fillna(42)
    b
w  42
x   8
y   8
z  42
```
Notice that the key column `a` is dropped. The index from original dataframe is kept. The row order of `fillna` columns matches that of the input. This is in contrast to a second representation from Pandas:
```python
>>> pdf = pd.DataFrame({'a': [3, 2, 3, 1], 'b': [None, 8, 8, None]}, index=['w', 'x', 'y', 'z'])
>>> pdf.groupby('a').fillna({'b': 42})
        b
a        
1 z  42.0
2 x   8.0
3 w  42.0
  y   8.0
```
Notice the rows are ordered by groups, not at their original input order. The index keeps both the keys column and the original index (reordered). Per dicsussion with @shwina, these two representations are interchangeable:
```python
>>> result = result.take(df['a'].argsort())
>>> result.index = cudf.DataFrame({'a': df['a'].sort_values(), None: result.index})
>>> result
      b
a      
1 z  42
2 x   8
3 w  42
  y   8
```
For now, in cudf we opt to support the first return format.

### CODE REVIEW

**Comment by ttnghia** [comment]
Did the name `fillna` exist in python ecosystem? If not then I think we should rename it to something else, because in C++ we have similar name `fill_n` (https://en.cppreference.com/w/cpp/algorithm/fill_n) which is very different. Using similar names but different meanings would lead to confusion/mistakes.

**Comment by shwina** [comment]
> Did the name fillna exist in python ecosystem? 

Yes: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html

**Review by devavret — APPROVED** [review]
Approving the one c++ change

File: cpp/src/groupby/groupby.cu [hunk 1]
```diff
@@ -246,8 +246,10 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> groupby::replace_nulls
                  std::back_inserter(results),
                  [&](auto i) {
                    auto grouped_values = helper().grouped_values(values.column(i), stream);
-                   return detail::group_replace_nulls(
-                     grouped_values->view(), group_labels, replace_policies[i], stream, mr);
+                   column_view view    = grouped_values->view();
+                   return view.nullable() ? detail::group_replace_nulls(
+                                              view, group_labels, replace_policies[i], stream, mr)
+                                          : std::move(grouped_values);
```
Comment (karthikeyann) [inline cpp/src/groupby/groupby.cu:252]: `grouped_values` is not allocated using `mr` allocator.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 2]
```diff
@@ -705,6 +707,114 @@ def cummax(self):
...
+        ----------
+        value : scalar, dict
+            Value to use to fill the holes. Cannot be specified with method.
+        method : {'backfill', 'bfill', 'pad', 'ffill', None}, default None
+            Method to use for filling holes in reindexed Series
+
+            - pad/ffill: propagate last valid observation forward to next valid
+            - backfill/bfill: use next valid observation to fill gap
+        axis : {0 or 'index', 1 or 'columns'}
+            Unsupoprted
```
Comment (karthikeyann) [inline python/cudf/cudf/core/groupby/groupby.py:773]:
```suggestion
            Unsupported
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]
C++ review

**Comment by isVoid** [comment]
Failed tests were caused by passing the wrong variable into the helper function, due to a typo. Will be more careful.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -760,6 +870,22 @@ def shift(self, periods=1, freq=None, axis=0, fill_value=None):
         result = self._groupby.shift(Table(value_columns), periods, fill_value)
         return self.obj.__class__._from_table(result)
 
+    def _mimic_pandas_order(
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/groupby/groupby.py:873]: If `_mimic_pandas_order` is expected to return DataFrame or Series does it mean that `fillna` might have chances of returning `DataFrame or Series` ?

Or the typing should have been `) -> DataFrame:`

File: python/cudf/cudf/core/groupby/groupby.py [hunk 4]
```diff
@@ -705,6 +707,114 @@ def cummax(self):
...
+            Unsupported
+        """
+        if limit is not None:
+            raise NotImplementedError("Does not support limit param yet.")
+
+        return self._scan_fill("bfill", limit)
+
+    bfill = backfill
+
+    def fillna(
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/groupby/groupby.py:752]: Should we also provide an option whether we need to follow pandas or not, and by default follow it ?

**Review by rgsl888prabhu — COMMENTED** [review]
Rest looks good.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 3 above)
Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:873]: Hmm, it should work with Series too. But when I'm testing it with Series, some `RuntimeError` occurs. Investigating.

Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:873]: Made a couple updates to `grouping`. Mainly to handle cases with specified with different `by` objects. The goal is to filter out the proper value columns that gets passed down to libcudf. Now `fillna` should handle `DataFrame` and `Series` just as fine.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 4 above)
Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:752]: Previous discussion with @shwina mentions that we probably don't want to introduce too many parameters to further confuse users. Especially at `groupby(sort=False)` level we already differ from Pandas.

Let that be said, `sort=False` doesn't seem to be oberseved by pandas for transform-like operation either:
```
>>> df = pd.DataFrame({'a': [3, 2, 3, 2], 'b': [5, None, None, 8]})
>>> df.groupby('a', sort=True).fillna(method='ffill')
     b
0  5.0
1  NaN
2  5.0
3  8.0
>>> df.groupby('a').fillna(method='ffill')
     b
0  5.0
1  NaN
2  5.0
3  8.0
```

cc @shwina

File: cpp/src/groupby/groupby.cu [hunk 5]
```diff
@@ -241,14 +241,19 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> groupby::replace_nulls
...
+  std::transform(
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(values.num_columns()),
+    std::back_inserter(results),
+    [&](auto i) {
+      bool nullable       = values.column(i).nullable();
+      auto final_mr       = nullable ? rmm::mr::get_current_device_resource() : mr;
+      auto grouped_values = helper().grouped_values(values.column(i), stream, final_mr);
+      return nullable ? detail::group_replace_nulls(
+                          *grouped_values, group_labels, replace_policies[i], stream, final_mr)
```
Comment (karthikeyann) [inline cpp/src/groupby/groupby.cu:254]: when nullable==true,`final_mr=rmm::mr::get_current_device_resource()`
so, returned value from `detail::group_replace_nulls` should get `mr`.
```suggestion
                          *grouped_values, group_labels, replace_policies[i], stream, mr)
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/src/groupby/groupby.cu (hunk 5 above)
Comment (isVoid) [reply cpp/src/groupby/groupby.cu:254]: oops, good catch.

File: python/cudf/cudf/tests/test_groupby.py [hunk 6]
```diff
@@ -1903,6 +1903,168 @@ def test_groupby_shift_row_zero_shift(nelem, fill_value):
...
+            {"dtype": "str", "null_frequency": 0.4, "cardinality": 10},
+        ],
+        rows=nelem,
+        use_threads=False,
+        seed=0,
+    )
+    key_col = "0"
+    value_cols = ["1", "2", "3", "4", "5"]
+    pdf = t.to_pandas()
+    gdf = cudf.from_pandas(pdf)
```
Comment (karthikeyann) [inline python/cudf/cudf/tests/test_groupby.py:1982]: why is this converted twice?

File: python/cudf/cudf/tests/test_groupby.py [hunk 7]
```diff
@@ -1903,6 +1903,168 @@ def test_groupby_shift_row_zero_shift(nelem, fill_value):
...
+            {"dtype": "str", "null_frequency": 0.4, "cardinality": 10},
+        ],
+        rows=nelem,
+        use_threads=False,
+        seed=0,
+    )
+    key_col = "0"
+    value_cols = ["1", "2", "3", "4", "5", "6"]
+    pdf = t.to_pandas()
+    gdf = cudf.from_pandas(pdf)
```
Comment (karthikeyann) [inline python/cudf/cudf/tests/test_groupby.py:1934]: why is this converted `to_pandas()` and then `from_pandas()`?

File: python/cudf/cudf/tests/test_groupby.py [hunk 8]
```diff
@@ -1903,6 +1903,168 @@ def test_groupby_shift_row_zero_shift(nelem, fill_value):
...
+            {"dtype": "str", "null_frequency": 0.4, "cardinality": 10},
+        ],
+        rows=nelem,
+        use_threads=False,
+        seed=0,
+    )
+    key_col = "0"
+    value_cols = ["1", "2", "3", "4", "5", "6", "7", "8"]
+    pdf = t.to_pandas()
+    gdf = cudf.from_pandas(pdf)
```
Comment (karthikeyann) [inline python/cudf/cudf/tests/test_groupby.py:2058]: same as above.

**Review by karthikeyann — COMMENTED** [review]
C++ looks good.
one question in pytest code.

File: python/cudf/cudf/tests/test_groupby.py (hunk 6 above)
Comment (isVoid) [reply python/cudf/cudf/tests/test_groupby.py:1982]: I believe `t` is a pyarrow table. `to_pandas` is a pyarrow function, not cudf's

**Review by karthikeyann — APPROVED** [review]
Look good.

**Comment by isVoid** [comment]
@gpucibot merge

**Pull request merged**